jitrebalance/jitrebalance.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+from pyln.client import Plugin, Millisatoshi
+from pprint import pprint
+from remote_pdb import RemotePdb
+from math import ceil
+import secrets
+import hashlib
+import binascii
+import time
+
+plugin = Plugin()
+pdb = RemotePdb('127.0.0.1', 4444)
+
+
+def get_circular_route(route, scid):
+    pass
+
+
+@plugin.async_hook("htlc_accepted")
+def on_htlc_accepted(htlc, onion, plugin, request, **kwargs):
+    pdb.set_trace()
+    plugin.log("Got an incoming HTLC htlc={}, onion={}".format(htlc, onion))
+
+    # The HTLC might be a rebalance we ourselves initiated, better check
+    # against the list of pending ones.
+    rebalance = plugin.rebalances.get(htlc['payment_hash'])
+    if rebalance is not None:
+        # Settle the rebalance, before settling the request that initiated the
+        # rebalance.
+        request.set_result({
+            "result": "resolve",
+            "payment_key": rebalance['payment_key']
+        })
+
+        # Now wait for it to settle correctly
+        # TODO Maybe be a bit smarter than having a fixed timeout here?
+        time.sleep(1)
+
+        rebalance['request'].set_result({"result": "continue"})
+        return
+
+    # Check to see if the next channel has sufficient capacity
+    scid = onion['per_hop_v0']['short_channel_id']
+
+    # Are we the destination? Then there's nothing to do. Continue.
+    if scid == '0x0x0':
+        request.set_result({"result": "continue"})
+        return
+
+    # Locate the channel + direction that would be the next in the path
+    peers = plugin.rpc.listpeers()['peers']
+
+    chan = None
+    peer = None
+    for p in peers:
+        for c in p['channels']:
+            if c['short_channel_id'] == scid:
+                chan = c
+                peer = p
+
+    # Check if the channel is active and routable, otherwise there's little
+    # point in even trying
+    if not peer['connected'] or chan['state'] != "CHANNELD_NORMAL":
+        request.set_result({"result": "continue"})
+        return
+
+    # Need to consider who the funder is, since they are paying the fees.
+    funder = chan['msatoshi_to_us_max'] == chan['msatoshi_total']
+    forward_amt = Millisatoshi(onion['per_hop_v0']['forward_amount'])
+    incoming_amt = Millisatoshi(htlc['amount'])
+    fee = incoming_amt - forward_amt
+    pprint(fee)
+    # Compute the amount we need to rebalance, give us a bit of breathing room
+    # while we're at it (25% more rebalancing than strictly necessary) so we
+    # don't end up with a completely unbalanced channel right away again, and
+    # to account for a bit of fuzziness when it comes to dipping into the
+    # reserve.
+    amt = ceil(int(forward_amt - chan['spendable_msat']) * 1.25)
```

Comment:
this will probably come with experience or by looking at past data of this channel. But I guess rebalancing the channel so that it is 50:50 afterwards might also be a plausible heuristic. 

you did not check in this case that your amt stays beyond 4.2M satoshi the max pay amt